build/v1/types.go

```diff
@@ -1022,23 +1026,37 @@ type ImageChangeTrigger struct {
+// generic ObjectReference BuildTriggerPolicy.ImageChange.From.  It is used by ImageChangeTriggerStatus, where a
+// specific instance of ImageChangeTriggerStatus in maintained in BuildConfigStatus.ImageChangeTriggers for each entry
+// in the BuildConfigSpec.Triggers array where the BuildTriggerPolicy.ImageChange pointer is set to a non-nil value
+type BuildImageChangeTriggerKey struct {
+	// namespace is the namespace where the ImageStreamTag used for an ImageChangeTrigger is located
+	Namespace string `json:"namespace,omitempty" protobuf:"bytes,1,opt,name=namespace"`
+
+	// name is the name of the ImageStreamTag used for an ImageChangeTrigger
+	Name string `json:"name,omitempty" protobuf:"bytes,2,opt,name=name"`
+}
+
 // ImageChangeTriggerStatus tracks the latest resolved status of the associated ImageChangeTrigger policy
 // specified in the BuildConfigSpec.Triggers struct.
 type ImageChangeTriggerStatus struct {
-	// lastTriggeredImageID is the sha/id of the imageref cited in the 'from' field the last time this BuildConfig was triggered.
+	// lastTriggeredImageID represents the current sha/id of the ImageStreamTag cited in the 'from' field of a corresponding
+	// ImageChangeTrigger, captured by the non-nil BuildTriggerPolicy.ImageChange pointer from a BuildConfigSpec.Triggers
+	// entry, the last time a Build for this BuildConfig was instantiated as a result of *ANY* of the ImageChangeTriggers
+	// in the BuildConfigSpec.Triggers array firing .
 	// It is not necessarily the sha/id of the image change that triggered a build.
-	// This field is updated for all image change triggers when any of them triggers a build.
+	// This field is updated for all image change triggers when any of them triggers a Build to prevent
+	// multiple builds being instantiated when multiple ImageChangeTriggers fire concurrently.
 	LastTriggeredImageID string `json:"lastTriggeredImageID,omitempty" protobuf:"bytes,1,opt,name=lastTriggeredImageID"`
 
 	// from is the ImageStreamTag that is used as the source of the trigger.
 	// This can come from an ImageStream tag referenced in this BuildConfig's triggers, or the From image in this BuildConfig's build strategy.
-	From *corev1.ObjectReference `json:"from,omitempty" protobuf:"bytes,2,opt,name=from"`
-
-	// paused is true if this trigger is temporarily disabled, and the setting on the spec has been reconciled. Optional.
-	Paused bool `json:"paused,omitempty" protobuf:"varint,3,opt,name=paused"`
+	From *BuildImageChangeTriggerKey `json:"from,omitempty" protobuf:"bytes,2,opt,name=from"`
 
-	// lastTriggerTime is the last time the BuildConfig was triggered by a change in the ImageStreamTag associated with this trigger.
-	LastTriggerTime metav1.Time `json:"lastTriggerTime,omitempty" protobuf:"bytes,4,opt,name=lastTriggerTime"`
+	// lastTriggerTime is the last time this particular ImageChangeTrigger fired and resulted in a Build being instantiated
```

Comment:
When combined with `lastTriggeredImageID` does this mean that the `lastTriggerTime` is the same for every entry in this slice?  Or does it mean that `lastTriggerTime` doesn't actually match the time `lastTriggeredImageID` value?

Reply:
lastTriggerTime == time at which *this* trigger lead to a build being run
lastTriggerImageID == the image id resolved by this trigger the last time the build was run for any reason  (probably not a great name, but trying to keep the values aligned w/ what we already have in spec).  The main purpose of this value is so we can determine "have we run a build since the last time the imagetag's SHA changed?  If not, we need to run one now and then update this field"

Reply:
lastly attempting to clarify this as well in next push